### PR TITLE
dnn : int8 quantized layers support in onnx importer

### PR DESCRIPTION
**Pull request by SamFC10**
**merge with extra** : https://github.com/opencv/opencv_extra/pull/896

Final PR for GSoC'21 project on 8-bit quantization support in dnn module. This PR adds new layers which are currently supported by onnx quantization.  
Docs - https://onnxruntime.ai/docs/how-to/quantization.html
Supported layers - https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/python/tools/quantization/registry.py#L34-L53

resolves :  #20188 
replaces #20264 

TODO:
- [x] tests for new layers
- [ ] ~~fallback to FP32 for unsupported cases~~ . Automatic fallback does not look possible
- [x] add support for unsupported cases - eltwise scalar input, concat layer, int8 resize layer
- [x] add quantized resnet50 onnx test - https://github.com/onnx/models/pull/460
- [ ] tutorial

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 1]
```diff
@@ -2427,6 +2502,378 @@ void ONNXImporter::parseCustom(LayerParams& layerParams, const opencv_onnx::Node
...
+        float out_sc = scale.at<float>(0);
+        int8_t out_zp = zeropoint.at<int8_t>(0);
+
+        Mat lookUpTable(1, 256, CV_8S);
+        int8_t* table = lookUpTable.ptr<int8_t>();
+        for (int i = -128; i < 128; i++)
+        {
+            float x = inp_sc*(i - inp_zp);
+            float y = x >= 0.f ? x : slope*x;
+            int quantized = out_zp + (int)std::round(y/out_sc);
```
[the remark this answers is not in the document]
Comment (SamFC10) [reply modules/dnn/src/onnx/onnx_importer.cpp]: Fixed

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 2]
```diff
@@ -469,6 +481,33 @@ void ONNXImporter::addConstant(const std::string& name, const Mat& blob)
     outShapes.insert(std::make_pair(name, shape(blob)));
 }
 
+void ONNXImporter::checkQuantizedLayer(const opencv_onnx::NodeProto& node_proto,
+                                       LayerParams& params, int& depth)
+{
+    // Quantized nodes have input and output names ending with 'quantized'
+    int len = node_proto.input(0).length();
```
[the remark this answers is not in the document]
Comment (SamFC10) [reply modules/dnn/src/onnx/onnx_importer.cpp]: Fixed.

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 3]
```diff
@@ -60,15 +60,18 @@ class ONNXImporter
 
     void addConstant(const std::string& name, const Mat& blob);
     void addLayer(LayerParams& layerParams,
-                  const opencv_onnx::NodeProto& node_proto);
+                  const opencv_onnx::NodeProto& node_proto, int depth = CV_32F);
```
[the remark this answers is not in the document]
Comment (SamFC10) [reply modules/dnn/src/onnx/onnx_importer.cpp:65]: You mean adding

```
layerParams.set("depth", CV_8S);
```

in all newly added nodes and then calling

```
layerParams.get<int>("depth");
```

inside addLayer() ?
Passing `depth` as an argument seems much cleaner to me as it is only used for allocating input/output blobs for that layer,  and not used by any layer's implementation.

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 4]
```diff
@@ -469,6 +481,33 @@ void ONNXImporter::addConstant(const std::string& name, const Mat& blob)
...
+        return;
+
+    if (node_proto.input(0).substr(len - 9) == "quantized")
+    {
+        depth = CV_8S;
+        std::string inpName = node_proto.input(0).substr(0, len - 9);
+        Mat scale = getBlob(inpName + "scale");
+        Mat zeropoint = getBlob(inpName + "zero_point");
+        params.set("scales", DictValue::arrayReal(scale.ptr<float>(), 1));
+        params.set("zeropoints", DictValue::arrayInt(zeropoint.ptr<int8_t>(), 1));
```
[the remark this answers is not in the document]
Comment (SamFC10) [reply modules/dnn/src/onnx/onnx_importer.cpp:523]: At present, only per-tensor scales and zeropoints for inputs/ouputs is supported by other frameworks (tflite, pytorch, onnx). Support for per-channel scales is limited to weights for convolution and fully-connected layer. So checking for per-tensor/per-channel scales in only one layer seemed enough to me.

File: modules/dnn/src/onnx/onnx_importer.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (alalek) [reply modules/dnn/src/onnx/onnx_importer.cpp:65]: Right.

`LayerParams` dictionary is designed for that. It is much flexible and less intrusive than approach with direct parameters. Keep changes minimal as possible, avoid copy-pasting code.


BTW, layers can utilize different types on outputs, e.g. #15996
The next, some backends / layers may not have int8 implementation or doesn't support some modes (you can process layers in importer, but you can't handle backends/targets in importer). So we need to have fallback on default FP32 calls for unsupported cases.

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 5]
```diff
@@ -2278,6 +2348,11 @@ void ONNXImporter::parseResize(LayerParams& layerParams, const opencv_onnx::Node
     for (int i = 1; i < node_proto.input_size(); i++)
         CV_Assert(layer_id.find(node_proto.input(i)) == layer_id.end());
 
+    int depth = CV_32F;
+    checkQuantizedLayer(node_proto, layerParams, depth);
+    if (depth == CV_8S)
+        CV_Error(Error::StsNotImplemented, "Int8 resize layer is not supported");
```
[the remark this answers is not in the document]
Comment (SamFC10) [reply modules/dnn/src/onnx/onnx_importer.cpp:2242]: Regarding fallback for current layers without int8 support and for new layers added in future :
1. Functions in 20228 can fallback to FP32 version automatically if int8 version of a layer is unavailable. Tests which have been added check the logic for the same. Adding int8 version of a new layer is optional.
2. Nodes in this PR which don't have int8 version donot fallback to FP32 version, they should be probably be added. Will try to add it in the coming days. Logic is slightly difficult as quantize/dequantize nodes have to be added before and after the unsupported layer.

File: modules/dnn/src/onnx/onnx_importer.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (SamFC10) [reply modules/dnn/src/onnx/onnx_importer.cpp]: Fixed in latest commit.

File: modules/dnn/src/onnx/onnx_importer.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (SamFC10) [reply modules/dnn/src/onnx/onnx_importer.cpp]: Fixed in latest commit

**Comment by SamFC10** [comment]
Marking pull request as ready for review as coding part is done. Only a short tutorial for using quantization and quantized onnx models remains.

@alalek Automatic FP32 fallback for an unsupported INT8 node does not look possible right now. Instead, adding INT8 path for that unsupported node seems much more easier. That's what I did with resize layer.

INT8 path for resize layer has been added in the FP32 version of the layer itself to avoid code duplication. But looking at it now, maybe the changes made are too much and may affect performance of the FP32 layer.  If required, I can move the int8 implementation to int8layers/resize_layer.cpp and revert the changes in layers/resize_layer.cpp.

**Comment by SamFC10** [comment]
Any solution to solve this build failure?

```
OpenCV tests: Can't find required data file: dnn/onnx/models/resnet50_int8.onnx in function 'findData'
" thrown in the test body
```

I have modified `download_models.py` in opencv/opencv_extra#896 to download quantized ResNet50 from onnx modelzoo.

**Comment by alalek** [comment]
`download_models.py` is not triggered automatically by CI. Now testdata share should be updated.

File: modules/dnn/test/test_onnx_importer.cpp [hunk 6]
```diff
@@ -967,6 +967,112 @@ TEST_P(Test_ONNX_layers, ConvResizePool1d)
     testONNXModels("conv_resize_pool_1d");
 }
 
+TEST_P(Test_ONNX_layers, Quantized_Convolution)
```
[the remark this answers is not in the document]
Comment (SamFC10) [reply modules/dnn/test/test_onnx_importer.cpp:970]: > This doesn't make sense for now

While it's true that other backends don't have quantized layer's implementation, we still need to ensure quantized networks fallback to a supported backend. Keeping the tests available for all backends helps in testing this fallback.

```
[ RUN      ] Test_ONNX_nets.ResNet50_Int8/0, where GetParam() = NGRAPH/CPU
[ WARN:0] global /home/jebastin/opencv_build/opencv/modules/dnn/src/dnn.cpp (4562) setPreferableBackend DNN: Only default backend supports quantized networks
FALLBACK: Layer [Quantize]:[data_quantized] is expected to has backend implementation
.
.
[       OK ] Test_ONNX_nets.ResNet50_Int8/0 (168 ms)
```
Internally, the code fallbacks to OpenCV/CPU target for unsupported backends, so I dont see a point in limiting the backends of these tests (except reducing the total time for running dnn module tests).

If you still think a new test fixture is needed, I will start working on it soon (will probably need some reference code)

**Review by alalek — APPROVED** [review]
Let put it in to have some usable tests for int8 feature.
Thank you :+1:

**Pull request merged**